### PR TITLE
feat(oval/suse): skip comparing TDC package and non-TDC package

### LINKED ISSUE #2024: False Positives in SUSE 12 SP3 Teradata kernels
# What did you do? (required. The issue will be **closed** when not provided.)
I scanned SUSE 12 SP3 with Teradata kernel installed (4.4.140-96.126.TDC.1) and we get hundreds of CVEs that is resolved in this version like CVE-2016-10905, vuls says that the fix version is 0:4.4.180-94.156.1 (not TDC) although the real fixed version is 4.4.140-96.97.TDC.1 as it appears in the oval file  https://ftp.suse.com/pub/projects/security/oval/suse.linux.enterprise.server.12.xml.gz
				<criterion test_ref="oval:org.opensuse.security:tst:2009876925" comment="kernel-default-4.4.140-96.97.TDC.1 is installed"/>
I verified that the CVE and proper fixed version are in the oval DB
kernel-default,0:4.4.121-92.169.1,5
kernel-default,0:4.4.180-94.156.1,6
kernel-default,0:4.4.140-96.97.TDC.1,6
so the bug is in vuls and how we query the suse oval DB for the kernel-default package.

# What did you expect to happen?
i expected not to see this CVE and many other CVEs

# What happened instead?
We got 100s of old/fixed/FP cves that are fixed inthe installed kernel-default

I think that for teradata packages (which are mainly kernel related packages) we should check the TDC entries in the DB and not the "regular" ones

### PR DESCRIPTION
**Pull request by MaineK00n**
If this Pull Request is work in progress, Add a prefix of “[WIP]” in the title.

# What did you implement:

Fixes #2024 

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)

# How Has This Been Tested?

by unit test

# Checklist:
You don't have to satisfy all of the following.

- [x] Write tests
- [ ] Write documentation
- [x] Check that there aren't other open pull requests for the same issue/feature
- [x] Format your source code by `make fmt`
- [x] Pass the test by `make test`
- [ ] Provide verification config / commands
- [x] Enable "Allow edits from maintainers" for this PR
- [x] Update the messages below

***Is this ready for review?:*** YES

# Reference

### CODE REVIEW

File: oval/util.go [hunk 1]
```diff
@@ -406,9 +406,15 @@ func isOvalDefAffected(def ovalmodels.Definition, req request, family, release s
...
-		// https://github.com/aquasecurity/trivy/blob/08cc14bd2171afdc1973c6d614dd0d1fb82b7623/pkg/detector/ospkg/oracle/oracle.go#L72-L77
-		if family == constant.Oracle && extractOracleKsplice(ovalPack.Version) != extractOracleKsplice(req.versionRelease) {
-			continue
+		switch family {
+		case constant.Oracle: // https://github.com/aquasecurity/trivy/blob/08cc14bd2171afdc1973c6d614dd0d1fb82b7623/pkg/detector/ospkg/oracle/oracle.go#L72-L77
+			if extractOracleKsplice(ovalPack.Version) != extractOracleKsplice(req.versionRelease) {
+				continue
+			}
+		case constant.SUSEEnterpriseServer:
+			if strings.Contains(ovalPack.Version, ".TDC.") != strings.Contains(req.versionRelease, ".TDC.") {
```
Comment (MaineK00n) [inline oval/util.go:415]: It would be preferable to have documentation from SUSE or TERADATA on how to handle this part.

**Comment by MaineK00n** [comment]
> @MaineK00n any work you still need to do with this PR? or it is ready to review and merge?

https://github.com/future-architect/vuls/pull/2025#discussion_r1765023170

File: oval/util.go (hunk 1 above)
Comment (shino) [reply oval/util.go:415]: I'm not sure but I think it's too rare that version string has ".TDC." when it's NOT teradata kernel.
So, +1 for this fix.

File: oval/util_test.go [hunk 2]
```diff
@@ -2364,6 +2364,68 @@ func TestIsOvalDefAffected(t *testing.T) {
...
+					},
+				},
+				req: request{
+					packName:       "kernel-default",
+					versionRelease: "0:4.4.140-96.97",
+					arch:           "x86_64",
+				},
+			},
+			affected: false,
+		},
```
Comment (shino) [inline oval/util_test.go:2407]: How about inversed pattern which is the very case of the corresponding issue? i.e.

```
		{
			in: in{
				family:  constant.SUSEEnterpriseServer,
				release: "12.3",
				def: ovalmodels.Definition{
					AffectedPacks: []ovalmodels.Package{
						{
							Name:    "kernel-default",
							Version: "0:4.4.180-94.156.1",
						},
					},
				},
				req: request{
					packName:       "kernel-default",
					versionRelease: "4.4.140-96.126.TDC.1",
					arch:           "x86_64",
				},
			},
			affected: false,
		},
```

**Review by shino — APPROVED** [review]
I added one comment but it's not mandatory.
🎊

File: oval/util_test.go (hunk 2 above)
Comment (MaineK00n) [reply oval/util_test.go:2407]: added
https://github.com/future-architect/vuls/compare/d942d938a2c182362eb516596dd2a83451cb13a2..eb9de339d3db209fd6dff8f44a00da9020ca4187

**Comment by wagde-orca** [comment]
@MaineK00n when do you pan to merge it?

**Pull request merged**